### ISSUE TITLE
Local and CI/CD isort with different behaviour 

### ISSUE THREAD

**Issue by matteoettam09** [PyCQA/isort#1889]
Hello,

have these lines in my .pre-commit.yaml:

```python
- repo: https://github.com/timothycrosley/isort
    rev: 5.9.3 # should be the same as in dev.in
    hooks:
      - id: isort
```

and locally I have version 5.9.3 of isort. When I run pre-commit run --all-files locally, the isort is passed succesfully, but when the pre-commit is performed in git, there are some differences:

```bash
$ pre-commit run --all-files
Check hooks apply to the repository......................................Passed
Check for useless excludes...............................................Passed
Check Yaml...............................................................Passed
Check for merge conflicts................................................Passed
Fix End of Files.........................................................Passed
Trim Trailing Whitespace.................................................Passed
prettier.................................................................Passed
isort....................................................................Failed
- hook id: isort
- files were modified by this hook
```

Locally I am in a venv, while remotely this is performed inside a docker image:

```bash
Running with gitlab-runner 14.6.0 (5316d4ac)
  on gitlab_runner_2 UfyZYyEg
Resolving secrets
00:00
Preparing the "docker" executor
00:03
Using Docker executor with image python:3.9 ...
Pulling docker image python:3.9 ...
Using docker image sha256:f88f0508dc467046a760e8ea0bb5c5861ff4e0ade96226ca4700abd1bd28b696 for python:3.9 with digest python@sha256:bf6f8b201ffac79a1b9fded783e7f0547ad2c2e3e48e96a7a23855c3de14df4b ...
```

Any idea why this inconsistency?
Thank you.

**Comment by danielgafni** [comment]
I have the same problem. isort suddenly started failing in my CI (in docker), while running it locally doesn't do anything. The version is 5.10.1, python 3.8.12.

**Comment by anirudnits** [comment]
Hey @danielgafni and @matteoettam09, this issue seems relevant to this earlier one: https://github.com/PyCQA/isort/issues/1790, I don't exactly know if this is applicable here but sharing if this gives any insight. In the meantime, could you provide me the link to the repo so that I could have a look?

**Comment by danielgafni** [comment]
I'm sorry, my project is private. 

I can only share my observations... the import line that was causing the issue was an import from the "scripts" directory. 

The project only had one import from this directory. All the other imports were either from standard/external libraries or from the project package. 

I hope this helps...

**Comment by matteoettam09** [comment]
Hi @anirudnits, thank you. My project is unfortunately private as well. 

In the meantime, I've figured it is possible to run the CI/CD locally, at least with Gitlab, so this could help identify the root cause.

**Comment by anirudnits** [comment]
> I can only share my observations... the import line that was causing the issue was an import from the "scripts" directory.

Can you provide the exact error? Like is this an issue where in the cicd pipeline the import from `scripts` directory is being classified as something other than firstparty?

@matteoettam09 Can you run `isort` with verbose option to identify exactly what the issue is?

~ Aniruddha

**Comment by danielgafni** [comment]
Yes! This happened to me too when I moved my .venv directory into my project dir!

**Comment by Magicbeanbuyer** [comment]
I encounter the same issue.

<img width="1316" alt="Screenshot 2022-08-25 at 12 22 16" src="https://user-images.githubusercontent.com/40656281/186640556-47520fc7-7294-4e4a-9995-65ce2ccfcf89.png">

I run isort view pre-commit both locally in docker container and in CircleCI container, both use the same docker image.

However locally isort behaves like the left side of the diff, and cicd the right side.

**Comment by maggielkx** [comment]
same problem encountered as @Magicbeanbuyer, i'm using python 3.8.11 and isrot 5.10.1 (with black profile)

**Comment by pgarrison** [comment]
I ran into the same issue and was able to [reproduce it here](https://github.com/pgarrison/isort-issue-1889-mwe).

In my case, the project has the following structure:
```
.
├── app
│   ├── pyproject.toml
│   └── util
│       └── constants.py
├── main.py
```

`pyproject.toml` is:
```
[tool.isort]
force_sort_within_sections = true
```

If I move `pyproject.toml` to the top level of the project structure, I am unable to reproduce the error. See the linked repository for a detailed log

**Comment by kashiwachen** [comment]
I got something similar. isort behaves differently at local with `isort --check--profile black .`. In addition, it behaves differently at project root and project sub-directory.
Here is my project structure:
```
.
├── Dockerfile
├── README.md
├── airflow
│   ├── airflow.cfg
│   ├── dags
│   │   ├── data_dag.py
```
If I run `isort --check --profile black ./` at root. It returns:
```sh
> Skipped 1 files
```
If I run `isort --check --profile black ./` at `dags` folder. It returns:
```sh
> ERROR: ./airflow/dags/data_dag.py Imports are incorrectly sorted and/or formatted.
```
And the changes from `isort --check --diff --profile black ./`:
```
 from datetime import datetime, timedelta

 import pendulum
-from utils.bash_util import gen_bash_command
-from utils.convert_util import A2B
-from utils.path_util import gen_path
-
 from airflow import models
 from airflow.operators.bash import BashOperator
 from airflow.operators.python import PythonOperator
 from airflow.providers.amazon.aws.hooks.s3 import S3Hook
+
+from utils.bash_util import gen_bash_command
+from utils.convert_util import A2B
+from utils.path_util import gen_path
```

**Comment by rafrafek** [comment]
`default_language_version` does not work because `isort` hook specifies language version. 

See comment: https://github.com/pre-commit/pre-commit/issues/2586#issuecomment-1305994077

**Comment by pgarrison** [comment]
@rafrafek do you mean that you expect your PR #1987 to fix this issue?

**Comment by rafrafek** [comment]
No, sorry for the confusion. I just found this issue by searching for `default_language_version`. I should create a new issue.

**Comment by pgarrison** [comment]
@anirudnits can we get the tags on this issue updated? I posted a MWE above, here's [the link again](https://github.com/pgarrison/isort-issue-1889-mwe)

**Comment by TheBeege** [comment]
This is affecting us, too. Has anyone isolated the issue?

**Comment by stinos** [comment]
Not sure if related, but I'm seeing things like

```
else-type place_module for foo returned FIRSTPARTY
else-type place_module for foo.scan returned THIRDPARTY
```

Now, I'm a first-time isort user so accompanying question: can I just 'merge' sections and tell isort to treat future/stdlib as one section, and have everything else go in another section? I don't really care about whether something is 'third' of 'first' anyway.

**Comment by Gornoka** [comment]
> Not sure if related, but I'm seeing things like
> 
> ```
> else-type place_module for foo returned FIRSTPARTY
> else-type place_module for foo.scan returned THIRDPARTY
> ```
> 
> I do have more than one foo directory if that matters, but only one is in the Python module search path, and that's the one which also has the scan subdirectory.
> 
> Now, I'm a first-time isort user so accompanying question: can I just 'merge' sections and tell isort to treat future/stdlib as one section, and have everything else go in another section? I don't really care about whether something is 'third' of 'first' anyway.


> As workaround I have added the following to my isort config.
> 
> ```haproxy-config
> [settings]
> known_first_party=my_package,my_package.something
> ```

**Comment by stinos** [comment]
@Gornoka yes I'm aware, but I'd really rather not start manually listing listing modules, that's not really manageable.

**Comment by jheddings** [comment]
I've seen the same behavior in [this job](https://github.com/jheddings/ansible-github/actions/runs/4255829522/jobs/7403937932).  Locally, `isort` separates the 3rd party groups, however the workflow groups them together.

In my case, my working copy was in folder named `github`, which is the same name as an external module.  Assumedly, this caused the external module to appear as local, which was obviously different in CI.  Once I changed the local folder name to something else (`ansible-github`), the CI action and my local action resulted in the same behavior.

So, in case others run into a similar issue, if your imports match the name of the current folder, you may have unexpected sorting results.

**Comment by fjoanr** [comment]
> I've seen the same behavior in [this job](https://github.com/jheddings/ansible-github/actions/runs/4255829522/jobs/7403937932). Locally, `isort` separates the 3rd party groups, however the workflow groups them together.
> 
> In my case, my working copy was in folder named `github`, which is the same name as an external module. Assumedly, this caused the external module to appear as local, which was obviously different in CI. Once I changed the local folder name to something else (`ansible-github`), the CI action and my local action resulted in the same behavior.
> 
> So, in case others run into a similar issue, if your imports match the name of the current folder, you may have unexpected sorting results.

I had the same issue, where isort would not change my files in local but then fail in the CI/CD job. Tested a lot of stuff, ensuring versions are the same, and even updated my Docker image for the runner. 

Turns out, we have a folder named "NLTK" within the repo, and the error was happening because `import nltk` would be moved around wrongly. The solution for me was to add :

```
known_third_party=nltk
```

to the `.isort.cfg`, and that did the trick. I guess another solution for us would be to rename `NLTK` to `nltk_utils` instead. But it is working now, so thanks for this thread!

**Comment by ed-randall-blk** [comment]
This same approach solved for me too.
We had a this import and the local vs CI/CD pipeline were fighting over its placement:
`from minio import Minio` 
I happened to have a subdir `minio` created during development to store a local minio db.   isort apparently considered that the name matched an import therefore it must be a first-party python package dir.  The pipeline on the other hand didn't have the `minio` subdir.
Renaming it to `minio-db` fixed the issue.

**Comment by fiendish** [comment]
Problem occurs for me because the pyproject.toml was located at a higher directory level than what I'm trying to isort, so it creates a found-local-dir vs installed-package conflict. My solution was moving the pyproject.toml to be next to the package I'm trying to sort imports in instead of up a directory.

**Comment by ctheune** [comment]
I'm also stumbling over this and it appears having multiple pyproject.toml files can confuse it when run from the CLI vs. from git pre-commit hooks.

So here: 
https://github.com/flyingcircusio/fc-nixos/tree/43990f24c03e4f25083bcb7bf1a4222dac86c3d2

We can see that our CI (which runs the pre-commit CLI) causes a change that puts 3rd party items into the local group:
https://github.com/flyingcircusio/fc-nixos/actions/runs/15007671114/job/42170173821?pr=1470

We have a top-level `pyproject.toml` file to unify the settings for Python files that are sprinkled in the repo and then we have `pyproject.toml` files down in the tree for actual packages. 

When I remove the topmost `pyproject.toml` file the CLI and the hook converge - but on the wrong decision! They will then not correctly identify the 3rd party packages and assume them to be local.

Our code is public and I can help figuring things out, but I don't have the isort code and details in my head's context window at the moment ... ;)

**Comment by pbasista** [comment]
What I *suspect* is happening in many of these confusing cases is that `isort` is invoked on multiple files at once. But it only uses a single configuration for all of them.

There are two main scenarios:

1. When `isort` is run by pre-commit for all files, e.g. via `pre-commit run isort --all-files`, it is invoked with all files in the git repository that are not supposed to be ignored by this hook as arguments. But it can only use one configuration. So it looks for a configuration suitable for the first Python file, finds it and then uses it for all the files in the repository.

This creates confusion when there are multiple different settings for `isort` in different parts of the repository. Or if some Python files in the repository do not have a specific configuration. Then the default one is used. For all files. The confusing part here is that the list of files passed to `isort` is not readily available to the user. So they cannot easily check which Python file was passed first and therefore which configuration is used. That is where many of the issues are coming from, in my opinion.

2. When `isort` is run for a subset of files, e.g. via `pre-commit run isort --files ...`, the same principle applies. The first Python file on the command line determines which configuration file will be used for all the other files. Here the user has some control over which files are checked and as long as the files of interest all use the same configuration, this command's behavior is fairly intuitive. It would do what the user asked it to do. Unless, of course, they prepend a Python file which uses a different configuration to the list.

`isort` has a command line option [`--resolve-all-configs`](https://pycqa.github.io/isort/docs/configuration/options.html#resolve-all-configs) which, if I understand correctly, should address this issue by using the closest local configuration for every processed file. But from what I have tried, that creates another issue.

The repository root is then considered to be the parent folder of all local packages. So, if your repository contains `subfolder/pyproject.toml` with `isort` settings defined and a local Python package `subfolder/project`, then `isort` would only consider `subfolder` to be the local package, rather than `project`. Which will have an effect of sorting imports from the `project` package, within the Python files in the `subfolder/project` directory, differently. They will be considered as 3rd party imports rather than local imports.

I am unsure if there currently is a reasonable way to configure `isort` to correctly handle processing in a single run of all files in a repository with multiple different Python packages in arbitrarily placed subdirectories that have different `isort` configurations.
